cmd/kubeadm/app/node/bootstrap.go

```diff
@@ -63,10 +104,10 @@ func EstablishMasterConnection(c *kubeadmapi.TokenDiscovery, clusterInfo *kubead
 	caCert := []byte(clusterInfo.CertificateAuthorities[0])
 
 	stopChan := make(chan struct{})
-	result := make(chan *ConnectionDetails)
+	result := make(chan *connectionDetails)
 	var wg sync.WaitGroup
 	for _, endpoint := range endpoints {
-		clientSet, err := createClients(caCert, endpoint, kubeadmutil.BearerToken(c), nodeName)
+		clientSet, err := createClients(caCert, endpoint, kubeadmutil.BearerToken(td), nodeName)
```

Comment:
I think createClients should return both the clientSet and the clientConfig (kubeconfig object).
If that particular clientSet can connect to the master, just return the clientConfig that was used.